crates/stages/src/stages/account_hashes.rs

```diff
@@ -0,0 +1,346 @@
+};
+use reth_db::{
+    cursor::{DbCursorRO, DbCursorRW},
+    database::Database,
+    models::AccountBeforeTx,
+    tables,
+    transaction::{DbTx, DbTxMut},
+};
+use reth_primitives::{keccak256, Address};
+use tracing::*;
+
+const ACCOUNT_HASHING: StageId = StageId("AccountHashing");
+
+/// The account hashing stage iterates over account states,
+/// hashes their addresses and inserts entries to the
+/// [`HashedAccount`][reth_interfaces::db::tables::HashedAccount] table.
+#[derive(Debug)]
+pub struct AccountHashingStage {
+    /// The size of inserted items after which the control
+    /// flow will be returned to the pipeline for commit
+    pub commit_threshold: u64,
+    /// The threshold for switching from incremental hashing
+    /// of changes to whole account hashing
+    pub clean_threshold: u64,
+}
+
+#[async_trait::async_trait]
+impl<DB: Database> Stage<DB> for AccountHashingStage {
+    /// Return the id of the stage
+    fn id(&self) -> StageId {
+        ACCOUNT_HASHING
+    }
+
+    async fn execute(
+        &mut self,
+        tx: &mut Transaction<'_, DB>,
+        input: ExecInput,
+    ) -> Result<ExecOutput, StageError> {
+        let previous_stage_progress = input.previous_stage_progress();
+        let stage_progress = input.stage_progress.unwrap_or_default();
+        let max_block_num = previous_stage_progress.min(stage_progress + self.commit_threshold);
+
+        // TODO: use macro here
+        if max_block_num <= stage_progress {
+            info!(target: "sync::stages::account_hashing", target = max_block_num, stage_progress, "Target block already reached");
+            return Ok(ExecOutput { stage_progress, done: true })
+        }
+
+        // We get the last transition from block `stage_progress` and add one to get the transition
+        // from the next block.
+        let start_transition = tx.get_block_transition_by_num(stage_progress)? + 1;
+        let end_transition = tx.get_block_transition_by_num(max_block_num)?;
+
+        if start_transition > end_transition {
+            // No transitions to walk over
+            info!(target: "sync::stages::account_hashing", start_transition, end_transition, "Target transition already reached");
+            return Ok(ExecOutput { stage_progress: max_block_num, done: true })
+        } else if end_transition - start_transition > self.clean_threshold || stage_progress == 0 {
+            // There are too many transitions, so we rehash all accounts
+            tx.clear::<tables::HashedAccount>()?;
+
+            let mut account_cursor = tx.cursor::<tables::PlainAccountState>()?;
+            let mut hashed_account_cursor = tx.cursor_mut::<tables::HashedAccount>()?;
+
+            let mut walker = account_cursor.walk(Address::zero())?;
+
+            while let Some((address, account)) = walker.next().transpose()? {
+                let hashed_address = keccak256(address);
+                hashed_account_cursor.append(hashed_address, account)?;
```

Comment:
Wanted to leave a comment as it will be valuable in future, this is Cursor footgun, if data is not presorted append will panic. `keccak256` will scramble the order. 
https://github.com/paradigmxyz/reth/blob/main/crates/storage/db/src/abstraction/cursor.rs#L79